cmd/perflock/perflock_test.go

```diff
@@ -0,0 +1,214 @@
+		t.Errorf("expected %d sleepers each sleeping %v to run sequentially, but time passed is %v",
+			len(sleepers), sleepDuration, got)
+	}
+}
+
+func TestShared(t *testing.T) {
+	t.Parallel()
+
+	socket := fmt.Sprintf("@perflock.%d.%s", os.Getpid(), funcname())
+
+	// 1. Start a daemon.
+	mustStartDaemon(t, socket)
+
+	// 2. Start three sleepers in SHARED mode, each sleeping for 0.5s.
+	start := time.Now()
+	var sleepers [3]*exec.Cmd
+	for i := range sleepers {
+		sleepers[i] = mustStartSleeper(t, socket, "--shared")
+	}
+
+	for _, sleeper := range sleepers {
+		sleeper.Wait()
+	}
+
+	// Assert that they ran concurrently by making sure it was shorter than than
+	// sleep_time*num_sleepers.
+	if got, maxTime := time.Since(start), time.Duration(len(sleepers))*sleepDuration; got > maxTime {
+		t.Errorf("expected %d shared sleepers each sleeping %v to not take as long as them sleeping sequentially, but time passed is %v",
+			len(sleepers), sleepDuration, got)
+	}
+}
+
+// funcname returns the function name of the caller.
+func funcname() string {
+	var pcs [1]uintptr
+	if runtime.Callers(2, pcs[:]) != 1 {
+		return "UNKNOWN"
+	}
+	fr, _ := runtime.CallersFrames(pcs[:]).Next()
+	return fr.Func.Name()
+}
+
+// mustStartSleeper starts a perflock client running a sleeper.
+func mustStartSleeper(t *testing.T, socket string, argv ...string) *exec.Cmd {
+	t.Helper()
+	cmd, err := startProcess(t, append(argv, "-socket="+socket, os.Args[0]), []string{"GO_TEST_MODE=perflock", "GO_TEST_PROGRAM_MODE=sleeper"})
+	if err != nil {
+		t.Fatalf("could not start sleeper: %v", err)
+	}
+	return cmd
+}
+
+// mustStartDaemon starts a perflock daemon and wait for it to start listening on
+// the socket.
+func mustStartDaemon(t *testing.T, socket string) {
+	t.Helper()
+	_, err := startProcess(t, []string{"-daemon", "-socket=" + socket}, []string{"GO_TEST_MODE=perflock"})
+	if err != nil {
+		t.Fatalf("could not start daemon: %v", err)
+	}
+	t.Logf("started daemon... waiting for it to become connectable")
+	{
+		ctx := context.Background()
+		ctx, cancel := context.WithTimeout(ctx, 3*time.Second)
+		defer cancel()
+		if !waitForDaemon(t, ctx, socket) {
+			t.Fatalf("gave up waiting for daemon")
+		}
+		t.Logf("daemon started!")
+	}
+}
+
+func startProcess(t *testing.T, argv []string, env []string) (*exec.Cmd, error) {
+	ctx := context.Background()
+	ctx, cancel := context.WithCancel(ctx)
+	cmd := exec.CommandContext(ctx, os.Args[0], argv...)
+	cmdReader, _ := cmd.StdoutPipe()
+	scanner := bufio.NewScanner(cmdReader)
+	var pid int
+	go func() {
+		envs := strings.Join(env, " ")
+		for scanner.Scan() {
+			fmt.Printf("[%12d] %-25s %s\n", pid, envs, scanner.Text())
+		}
+	}()
+
+	cmd.Stderr = cmd.Stdout
+	cmd.Env = os.Environ()
+	cmd.Env = append(cmd.Env, env...)
+	if err := cmd.Start(); err != nil {
+		cancel()
+		return nil, err
+	}
+	pid = cmd.Process.Pid
+	t.Cleanup(func() {
+		cancel()
+		if err := cmd.Wait(); err != nil {
```

Comment:
Just for robustness, set cmd.WaitDelay?

Reply:
I'm developing this on a Debian stable (Crostini on a Chromebook), so the version is 1.19 and WaitDelay was introduced in 1.20. I've now upgraded my toolchain to 1.21 so I can use the latest and greatest stuff. I've updated the go.mod file for the new minimum version.